src/libraries/System.Text.Json/src/System/Text/Json/Serialization/ReflectionEmitMemberAccessor.cs

```diff
@@ -262,6 +270,9 @@ private static DynamicMethod CreatePropertySetter(PropertyInfo propertyInfo, Typ
             Type? declaringType = propertyInfo.DeclaringType;
             Debug.Assert(declaringType != null);
 
+            Type realPropertyType = propertyInfo.PropertyType;
+            Debug.Assert(propertyType.IsAssignableFrom(realPropertyType));
```

Comment:
Since in case of type inequality there will be emitted a cast instruction, there is no need in that line. It just partly duplicates logic of `unbox_any`.

Reply:
I like assertions for clearness

Reply:
Assertions are okay, but the runtime check won't hurt in any case since it will happen only when jitting.